File: unity_kit/ios/Classes/UnityEventListener.swift
import Foundation

/// Protocol for receiving Unity player events.
///
/// Implemented by view controllers that need to react to Unity lifecycle
/// and messaging events. Uses `@objc` for Objective-C runtime compatibility
/// with UnityFramework callbacks.
@objc protocol UnityEventListener: AnyObject {

    /// Called when Unity sends a message to Flutter.
    ///
    /// - Parameter message: Raw message string from Unity's
    ///   `UnityMessageManager.Instance.SendMessageToFlutter()`.
    func onMessage(_ message: String)

    /// Called when a Unity scene finishes loading.
    ///
    /// - Parameters:
    ///   - name: Scene name.
    ///   - buildIndex: Scene build index.
    ///   - isLoaded: Whether the scene is loaded.
    ///   - isValid: Whether the scene reference is valid.
    func onSceneLoaded(_ name: String, buildIndex: Int32, isLoaded: Bool, isValid: Bool)

    /// Called when the Unity player is created and ready.
    func onCreated()

    /// Called when the Unity player has been unloaded.
    func onUnloaded()
}

File: unity_kit/android/src/main/kotlin/com/unity_kit/UnityEventListener.kt
package com.unity_kit

/// Listener interface for Unity player events.
///
/// Implementations receive callbacks when the Unity player emits events
/// such as messages, scene loads, creation, and unloading.
interface UnityEventListener {
    /// Called when Unity sends a message to Flutter.
    fun onMessage(message: String)

    /// Called when a Unity scene finishes loading.
    fun onSceneLoaded(name: String, buildIndex: Int, isLoaded: Boolean, isValid: Boolean)

    /// Called when the Unity player is created and ready.
    fun onCreated()

    /// Called when the Unity player is unloaded.
    fun onUnloaded()
}
